src/client/marble-skins.js

```diff
@@ -0,0 +1,175 @@
+											}
+
+											return undefined;
+										})
+								);
+							}
+
+							return Promise.all(shaderFilePromises)
+								.then((values) => {
+									// Copy new uniforms into live-updating object.
+									if (skinMeta.textures) {
+										let newUniforms = {};
+										for (let key in skinMeta.textures) {
+											newUniforms[key] = { value: _textureLoader.load(`resources/skins/${skinId}/${skinMeta.textures[key]}`) };
+											newUniforms[key].value.wrapS = newUniforms[key].value.wrapT = THREE.RepeatWrapping;
+										}
+										Object.assign(_shaderUniforms, newUniforms);
+									}
+
+									// Create the material
+									_skins[skinId].material = new THREE.ShaderMaterial({
+										fragmentShader: values[0],
+										vertexShader: values[1],
+										uniforms: _shaderUniforms
+									});
+
+									renderCore.shaderMaterials.push(_skins[skinId].material);
+
+									return _skins[skinId].material;
+								});
+						} else {
+							// Load various maps. This tries to load all map types. If loading fails (because
+							// it doesn't exist), it will simply ignore that map type.
+							let mapTypes = ["diffuse", "normal"];
+							let mapTexturePromises = [];
+
+							// Load every map type as a texture. The texture loader has been promisified.
+							for (let map of mapTypes) {
+								mapTexturePromises.push(
+									new Promise((resolve) => {
+										// Load texture
+										_textureLoader.load(
+											`resources/skins/${skinId}/${map}.png`,
+											(texture) => {
+												_skins[skinId][map] = texture;
+												resolve(texture);
+											},
+											undefined,
+											() => {
+												resolve(map === "diffuse" ? _fallbackDiffuse : undefined);
+											}
+										);
+									})
+								);
+							}
+
+							// After all texture maps are loaded, create the material
+							return Promise.all(mapTexturePromises)
+								.then(() => {
+									let materialValues = {};
+									materialValues.color = new THREE.Color(color || "#ffffff");
+									materialValues.roughness = !isNaN(skinMeta.roughness) ? skinMeta.roughness : .9;
+									materialValues.metalness = !isNaN(skinMeta.metalness) ? skinMeta.metalness : 0;
+									materialValues.map = _skins[skinId].diffuse;
+									if (_skins[skinId].normal) {
+										materialValues.normalMap = _skins[skinId].normal;
+									}
+
+									return _skins[skinId].material = new THREE.MeshStandardMaterial(materialValues);
+								});
+						}
+					})
+					.catch(() => {
+						// Skin failed to load, so remove it from the skins list
+						delete _skins[skinId];
+					});
+			} else {
+				// Skin meta has already been retrieved before, so we only have to return it. Since the
+				// return value is always expected to be a promise, we will return a promise that
+				// immediately resolves to the already loaded material.
+				return Promise.resolve(_skins[skinId].material);
```

Comment:
Because the creation of the material also sets the colour, any marbles that spawn with the same skin have the colour of the first marble that spawned with it.
If you spawn multiple marbles at once (or refresh the client) the colours appear to be fixed, but that's because all the promises are being resolved at the same time, and those skins aren't re-used at that point (meaning we're loading the same skin multiple times when that happens).

We may have to store the loaded shaders/textures per skin instead, and then generate materials from them.

Reply:
Ah yes, the custom color thing was pretty much added in the last, as an afterthought. I kind of want to re-do it so that we're not stuck with the default blend mode for the rest of the textures.

Regarding how to store the skins, we might want to do that only for the skins that have custom color enabled.

I'll think about this some more.

Reply:
The colour issue is resolved, but it's technically still bugged in the way that it will load the same skin multiple times if multiple marbles spawn in at once.

In the end they all load the same thing and overwrite each other in `_skins[skinId]` meaning there's no lasting issue, but letting the client load the same skin multiple times doesn't look very good still. Perhaps you can resolve it by storing the skins as Promises with the actual skin Object as it's resolved return value.